### ISSUE TITLE
JsonParseException exception thrown on cljr-add-project-dependency

### ISSUE THREAD

**Issue by mbuczko** [clojure-emacs/refactor-nrepl#226]
When calling `cljr-add-project-dependency` from clj-refactor it ends up with exception being thrown from refactor-nrepl:

```
Caused by: refactornrepl240SNAPSHOT.com.fasterxml.jackson.core.JsonParseException: Unexpected character (’<’ (code 60)): expected a valid value (number, String, array, object, ’true’, ’false’ or ’null’)
 at [Source: (StringReader); line: 1, column: 2]
 at refactornrepl240SNAPSHOT.com.fasterxml.jackson.core.JsonParser._constructError (JsonParser.java:1798)
    refactornrepl240SNAPSHOT.com.fasterxml.jackson.core.base.ParserMinimalBase._reportError (ParserMinimalBase.java:663)
    refactornrepl240SNAPSHOT.com.fasterxml.jackson.core.base.ParserMinimalBase._reportUnexpectedChar (ParserMinimalBase.java:561)
    refactornrepl240SNAPSHOT.com.fasterxml.jackson.core.json.ReaderBasedJsonParser._handleOddValue (ReaderBasedJsonParser.java:1892)
    refactornrepl240SNAPSHOT.com.fasterxml.jackson.core.json.ReaderBasedJsonParser.nextToken (ReaderBasedJsonParser.java:747)
    mranderson048.cheshire.v5v8v0.cheshire.parse$parse.invokeStatic (parse.clj:88)
    mranderson048.cheshire.v5v8v0.cheshire.parse$parse.invoke (parse.clj:86)
    mranderson048.cheshire.v5v8v0.cheshire.core$parse_string.invokeStatic (core.clj:206)
    mranderson048.cheshire.v5v8v0.cheshire.core$parse_string.invoke (core.clj:192)
    mranderson048.cheshire.v5v8v0.cheshire.core$parse_string.invokeStatic (core.clj:203)
    mranderson048.cheshire.v5v8v0.cheshire.core$parse_string.invoke (core.clj:192)
    refactor_nrepl.artifacts$get_versions_BANG_.invokeStatic (artifacts.clj:72)
    refactor_nrepl.artifacts$get_versions_BANG_.invoke (artifacts.clj:62)
    refactor_nrepl.artifacts$collate_artifact_and_versions.invokeStatic (artifacts.clj:80)
    refactor_nrepl.artifacts$collate_artifact_and_versions.invoke (artifacts.clj:78)
    clojure.core$partial$fn__5561.invoke (core.clj:2616)
    clojure.core$map$fn__5587.invoke (core.clj:2747)
    clojure.lang.LazySeq.sval (LazySeq.java:40)
    clojure.lang.LazySeq.seq (LazySeq.java:49)
    clojure.lang.RT.seq (RT.java:528)
    clojure.core$seq__5124.invokeStatic (core.clj:137)
    clojure.core$map$fn__5587.invoke (core.clj:2738)
    clojure.lang.LazySeq.sval (LazySeq.java:40)
    clojure.lang.LazySeq.seq (LazySeq.java:49)
    clojure.lang.Cons.next (Cons.java:39)
    clojure.lang.RT.next (RT.java:706)
    clojure.core$next__5108.invokeStatic (core.clj:64)
    clojure.core$dorun.invokeStatic (core.clj:3134)
    clojure.core$dorun.invoke (core.clj:3125)
    refactor_nrepl.artifacts$add_artifacts$fn__54743$fn__54744.invoke (artifacts.clj:89)
    clojure.core$binding_conveyor_fn$fn__5476.invoke (core.clj:2022)
    clojure.lang.AFn.call (AFn.java:18)
    java.util.concurrent.FutureTask.run (FutureTask.java:266)
    java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1142)
    java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:617)
    java.lang.Thread.run (Thread.java:745)
```

**Comment by mbuczko** [comment]
looks like it's a problem of http://search.maven.org, which throttles requests and after 32 request (or so..) returns:

```
HTTP/1.1 503 Service Temporarily Unavailable
Content-Type: text/html
Date: Fri, 22 Jun 2018 12:46:46 GMT
Server: nginx/1.12.1
Content-Length: 213
Connection: Close

<html>
<head><title>503 Service Temporarily Unavailable</title></head>
<body bgcolor="white">
<center><h1>503 Service Temporarily Unavailable</h1></center>
<hr><center>nginx/1.12.1</center>
</body>
</html>
```

**Comment by benedekfazekas** [comment]
hm.. thanks for digging into this. i've seen this the other day but assumed something with the corporate proxy i was behind..

**Comment by mbuczko** [comment]
I thought the same at the beginning. Now, I'm going with no proxy and this problem occurs every single time, always after 32nd request which, when fired few seconds after, returns valid response.

To be precise, this is a request for artifact versions, so maybe it would be worth to decouple "get versions" requests from "get list of artifacts" ones and fire former only after particular artifact has been already selected? That would reduce number of "get versions" request just to one.

**Comment by expez** [comment]
Yeah, seems we have to do that, given this new ratelimit.

**Comment by eigenhombre** [comment]
I also have this problem.

**Comment by mbuczko** [comment]
I wonder if we could not change user-experience a bit and solve https://github.com/clojure-emacs/clj-refactor.el/issues/413 by the way. As for now, we fetch all the artifacts at once which has one advantage - filtering is instantaneous. On the other hand, if we fetched artifacts as user types them (assuming some kind of debouncing to not generate too many requests) we will introduce a bit of latency but: 1. we won't generate as many requests at once. 2. we could list and fetch actually every artifact from mvn repository, without limiting ourselves to "org.clojure" and "com.cognitect" groups.

I implemented very similar mechanism in https://github.com/mbuczko/helm-clojuredocs, so if you think it makes sense to have something like this here, I'm happy to take it over.

**Comment by expez** [comment]
@mbuczko That sounds great.  We just have to make sure you can find what you're looking for by entering e.g. `friend` without having to remember that it resides under `com.cemerick`

**Comment by vxe** [comment]
additionally why are calls to `cljr-hotload-dependency` , failing in the same manner?  I would expect the rate limit to be a non issue in this case

**Comment by expez** [comment]
It might be doing something quite stupid.

This code hasn't really been revisited since it was first written.  Could definitely use some love! :)

**Comment by expez** [comment]
@mbuczko are you working on this?  We really need to cut a new release and this fix should be included.

**Comment by mbuczko** [comment]
@expez sorry, I've just came back from a week of family vacations. 

I would need about 3 days to implement a fix, so if you want to make a release within day or two I don't think I will be able create PR that fast.

**Comment by expez** [comment]
@mbuczko Some time this week would be fine :)

**Comment by mbuczko** [comment]
ok, so implementing a _load-artifacts-as-you-type_ mechanism as I described before seems to be more challanging than I expected. we use `completing-read` in clj-refactor which does not allow easy debouncing of multiple keystrokes in a very short time. Helm does it out of the box, but moving all its machinery here is a bit of overkill.

Instead, I fixed the way how we fetch versions, which effectively solves the problem of throttled requests. Now, maven-originated artifacts are not scanned upfront for their versions, but a separate request is fired once a maven-based artifact has been chosen.
